### ISSUE TITLE
Editing a commit with an interactive rebase is not properly handled

### ISSUE THREAD

**Issue by AdrienMatricon** [sourcegit-scm/sourcegit#790]
Steps to reproduce:
- Perform an interactive rebase
- Choose Edit for a commit
- Click on `Start`

What that does:
- Good: It does perform an interactive rebase and stops at that commit so that the user can edit it
- Good: There is a banner at the top saying that there is an ongoing rebase
- Bad: The buttons in the banner are `Resolve` and `Abort` as if there was a conflict
  * There is no conflict to resolve, so the buttons should be `Continue` and `Abort`
  * I may be remembering this incorrectly but I think that this used to be handled properly until some weeks ago
- Bad: In the `Local Changes` page, the `Amend` checkbox has disappeared, and the `Commit` and `Commit & Push` buttons are replaced by a `Continue` button and a menu with the option `Commit (edit)`
  * Editing is supposed to be like working on a branch, except that you cannot rebase because there is already a rebase going on. Some example of its uses: amending the edited commit, creating a new commit after the edited one, splitting the edited commit (and maybe also some before it) by resetting to an older commit then creating better commits with those changes. Then once you're done, you tell the rebase to continue so that it applies the later commits on top of the changes
  * For that reason, the page should be as usual, allowing the user to stage/commit/amend
  * That being said, removing the `Commit & Push` button does make sense because pushing during a rebase is an error (although you could keep it and let the user deal with the error message), and renaming the `Commit` button to `Commit (edit)` is not a bad idea

**Comment by love-linger** [comment]
#773
#767
#693

**Comment by Cyrillev91** [comment]
Hi all,

> Bad: The buttons in the banner are Resolve and Abort as if there was a conflict
>     There is no conflict to resolve, so the buttons should be Continue and Abort
>     I may be remembering this incorrectly but I think that this used to be handled properly until some weeks ag

The role of the `Resolve` button is to go to the `Local changes` page
- In the case of a rebase it is to `resolve` a conflict
- In the case of an interactive rebase it is to edit a commit (maybe also resolve a conflict). In this case the button does not necessarily have a clear name.

note 1: This Resolve button has no equivalent in git command
nota 2: If it was renamed to go to Local Changes, I don't know if that would be meaningful and if it is deleted I don't think it would be better




> the Amend checkbox has disappeared

This was fixed after #773
However, it should be checked in the case of 'Edit' (interactive resabe) as explained #773

```
I think that for a better user experience:
We should display the commit marked with edit in the same way as when the amend button is checked in the case of the rewriting of the last commit:

    - check the Amend box to clearly identify that we are rewriting a commit (It is not possible to do it manually because the checkbox is not displayed)
    - in the stage: display the current files of the commit
    - display the comment of the edited commit
    - the commit button could be called "Ammend Commit" for clarity.
    So the user can modify the commit marked edit :
    - add a forgotten file (with the sourcegit GUI)
    - modify a file (with the sourcegit GUI)
    - delete a file (with the sourcegit GUI)
    - modify the commit message.
    - Validate the changes of this commit by pressing the "Amend Commit" button
    (possibly add a new commit as suggested in #767.
Then click on the continue button to continue the rebase.
```


> Bad: The buttons in the banner are Resolve and Abort as if there was a conflict
>      There is no conflict to resolve, so the buttons should be Continue and Abort

If the continue button had to return to the banner it would only have to be placed on the Local changes page, and not on the History page where you have to keep the RESOLVE (the one that goes to the local changes page).

I think the continue button can be displayed instead of the commit button (to be consistent with the Rebase (without interactive) see #693.
After the discussion is open, and constructive criticism is welcome.



>     Editing is supposed to be like working on a branch, except that you cannot rebase because there is already a rebase going on. Some example of its uses: amending the edited commit, creating a new commit after the edited one, splitting the edited commit (and maybe also some before it) by resetting to an older commit then creating better commits with those changes. Then once you're done, you tell the rebase to continue so that it applies the later commits on top of the changes
>     For that reason, the page should be as usual, allowing the user to stage/commit/amend
>     That being said, removing the Commit & Push button does make sense because pushing during a rebase is an error (although you could keep it and let the user deal with the error message), and renaming the Commit button to Commit (edit) is not a bad idea
> 
I agree.

**Comment by AdrienMatricon** [comment]
> The role of the Resolve button is to go to the Local changes page
> ...
> Note 1: This Resolve button has no equivalent in git command

I must admit that there was no doubt in my mind that `Resolve` was equivalent to the git command `git mergetool`. I'm not sure there needs to be a button to go to local changes when one can already click on "Local changes" directly

Here is what I suggest in terms of interface/ergonomy:
- I think the banner is a great way to show that you are in a special state and that there specific actions that you can take because of that
- For that reason, I think `Continue` and `Abort` should always both be in the banner, maybe with a button for more options (i.e. for the actions that can be taken during merge/rebase/cherry-pick, like  `--skip`, `--quit` or `--edit-todo`)
- I think a button to open the mergetool would be good as well if there isn't one already (maybe repurpose `Resolve` to do that?), but only if there is an actual conflict to resolve
- I understand that when solving a conflict, you're supposed to stage the version you want then run `--continue` instead of performing a commit, so it makes sense to turn the `Commit` button into a `Continue` button in that case. This should be based on whether there's a conflict, not whether there's a rebase going on.
- I'm less sure about redirecting people to the `Local Changes` page when they have already solved the conflict and just wish to proceed, but if the point is that people double-check what they've staged then maybe the `Continue` button in the banner should be the one to redirect to that page (and then work normally or be grayed out if you're already on that page)

**Comment by love-linger** [comment]
* When there are no conflicts and no editing commits, why do you think the merge/rebase/cherry-pick/revert action will stop to wait user clicking the `--continue` button?
* When the merge/rebase/cherry-pick/revert action stops and is waiting for user resolving conflicts or editing current commit, if we do NOT redirect user to the `Local Changes`, how do we let users know which file has conflicts? Or is the `Local Changes` page not the right page user editing commits?
* For `interactive rebase`, if you choose multiple commits with `edit`, and when the rebasing goes to one of them, it will stop and wait for user editing the current one. If we do NOT take users to the `LOCAL CHANGES`, how do we let users known which commit is current processing? BTW, in this condition, the `continue` button is enabled because there's no conflicts.

**Comment by love-linger** [comment]
I've pushed some modifications to `develop`.

![image](https://github.com/user-attachments/assets/d3e62343-6322-471a-bb99-a8c76ab6f70c)

* Show the current head commit for `cherry-pick`
* Show the stopped at commit for `rebase`
* Add a `SKIP` button for `cherry-pick` and `rebase`

**Comment by AdrienMatricon** [comment]
> When there are no conflicts and no editing commits, why do you think the merge/rebase/cherry-pick/revert action will stop to wait user clicking the --continue button?

I don't. I think that pretty much covers the cases in which git will pause in the middle of the command and leave you in a special context.

> When the merge/rebase/cherry-pick/revert action stops and is waiting for user resolving conflicts or editing current commit, if we do NOT redirect user to the Local Changes, how do we let users know which file has conflicts? Or is the Local Changes page not the right page user editing commits?

- First, I think the banner should explicitly say that there is a conflict to fix when there is one.
- Beyond that... I don't think anything else actually needs to be done here. If the user knows how to solve conflicts, they know what they need to do and can go to the `Local Changes` page without a dedicated button in the banner (which is confusing when you don't know what the button actually does before clicking it). If they don't, I'm not even sure they would really know what to do on the `Local Changes` page.
- That being said, maybe the `Local Changes` button could be highlighted in some way when there is a conflict going on (maybe add a small red rectangle on the side with "conflict resolution" written on it?)
- On the `Local Changes` page, the conflicting files could stand out more, though. When running `git status` they are listed separately as "both modified". In `Local Changes` I think there's an icon to tell you those are conflicting files, but that is not much. You could put them at the top with a red background or something, though I believe listing them separately from staged/unstaged files would be best, as they are different: you don't see the changes on the right when selecting them, and the actions you can perform on them are not the same

> For interactive rebase, if you choose multiple commits with edit, and when the rebasing goes to one of them, it will stop and wait for user editing the current one. If we do NOT take users to the LOCAL CHANGES, how do we let users known which commit is current processing? BTW, in this condition, the continue button is enabled because there's no conflicts.

Well, you're not really lost because `HEAD` is visibly pointing to the commit that you are editing, but I agree that isn't very user friendly.

If we go back to the git command line:
- when you run `git rebase`, it tells you:
```
Stopped at 7ed342eb...  My commit message
You can amend the commit now, with
  git commit --amend 
Once you are satisfied with your changes, run
  git rebase --continue
```
- when you run `git status`, it tells you:
```
Last commands done (7 commands done):
   pick c58f58af Some commit message
   edit 7ed342eb Other commit message
  (see more in file .git/rebase-merge/done)
Next commands to do (6 remaining commands):
   pick fcc6ba30 Yet another commit message
   pick fbe6cf55 Still another one
  (use "git rebase --edit-todo" to view and edit)
You are currently editing a commit while rebasing branch 'myBranch' on '11f5cb36'.
  (use "git commit --amend" to amend the current commit)
  (use "git rebase --continue" once you are satisfied with your changes)
```

Similarly, you could:
- Show what commit is in progress in the banner like you already did for `cherry-pick` (but maybe also show the commit message's subject if there's enough room or as a hover text)
- Have a `See Command List` or `See More` button as the first one in the banner to display where we are within the interactive rebase through a window similar to the interactive-rebase one. You could even make future commands editable to integrate the `--edit-todo` feature

**Comment by leo-liar** [comment]
If I may add the following issue while doing an interactive rebase: (_btw thanks for implementing that!_)

I want to reword some commit messages and when opening the edit field via the small pen icon then the whole previous commit message is squashed into the subject line.  The screenshot shows what is seen immediately after opening the edit field:

![Interactive_Rebase](https://github.com/user-attachments/assets/7cf9596a-4e77-4ebf-9c71-09bf552c9905)

A bit of experimenting shows, that this is due to SourceGit not handling `LF` (`\n`) only ending lines in these cases. E.g. when I'm writing a new commit message (with subject and body separated by blank line) in some external editor and ensure `CR` `LF` (`\r\n`) line breaks, then SourceGit automatically detects this and fills each field accordingly. When using only `LF` then not.

(Quiete strange too is that the old commit message was also written in SourceGit and while being messed up in the edit field is correctly previewed at the bottom.)

(I'm using SourceGit on Win 10.)

**Comment by love-linger** [comment]
> If I may add the following issue while doing an interactive rebase: (_btw thanks for implementing that!_)
> 
> I want to reword some commit messages and when opening the edit field via the small pen icon then the whole previous commit message is squashed into the subject line. The screenshot shows what is seen immediately after opening the edit field:
> 
> ![Interactive_Rebase](https://private-user-images.githubusercontent.com/25538015/395124674-7cf9596a-4e77-4ebf-9c71-09bf552c9905.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwMDI5MzMsIm5iZiI6MTczNDAwMjYzMywicGF0aCI6Ii8yNTUzODAxNS8zOTUxMjQ2NzQtN2NmOTU5NmEtNGU3Ny00ZWJmLTljNzEtMDliZjU1MmM5OTA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEyVDExMjM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhZDA3ZGUwOTdlOTE1ZTMwYWQ5NjA4MTEzNGFkZWI1Yzg2ZGQwYjdjZjQ5NTBiMjJkNjhmY2Y0OGMxNzEzYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uGQ1pUys3rgLlsBj_LjCsBiXLjL46SIpLhgPTY23tGU)
> 
> A bit of experimenting shows, that this is due to SourceGit not handling `LF` (`\n`) only ending lines in these cases. E.g. when I'm writing a new commit message (with subject and body separated by blank line) in some external editor and ensure `CR` `LF` (`\r\n`) line breaks, then SourceGit automatically detects this and fills each field accordingly. When using only `LF` then not.
> 
> (Quiete strange too is that the old commit message was also written in SourceGit and while being messed up in the edit field is correctly previewed at the bottom.)
> 
> (I'm using SourceGit on Win 10.)

Fixed. You can download the latest CI build from Github Action.

**Comment by leo-liar** [comment]
Another line break error:

The shown error message(s) (in thise case triggered by aforementioned hook failing) do also remove _all_ blank lines (and those only containing whitespaces).
![image](https://github.com/user-attachments/assets/6962d22d-6bee-43ea-b2be-6e0664a6c30f)

On the command line commit messages are shown as entered (i.e. with blank lines).

**Comment by love-linger** [comment]
The `Trim()` in 59f11e9acaf5edbf6dad64565dd550e031f4aa67 and 6eb226eb4492d2fe6a05fb3162da0f782d3f5fc1 is only used for reading and displaying commit message (not writing).

**Comment by love-linger** [comment]
Most of the error messages are reading from `Process.ErrorDataReceived` event handler, which will triggerer when there is a new line receviced from `stderr`, and if it is an empty string (blank line), `SourceGit` will just ignore it.

**Comment by leo-liar** [comment]
> Most of the error messages are reading from `Process.ErrorDataReceived` event handler, which will triggerer when there is a new line receviced from `stderr`, and if it is an empty string (blank line), `SourceGit` will just ignore it.

Ah, I see. For the sake of legibility it might be better to not strip those blank lines, imho. And if I might add a whish – it would be nice to be able to resize the error message window.

**Comment by love-linger** [comment]
If your whitespace is in start of subject or endof description, it will be trimmed. Otherwise, all the content will be writed into a temp file, then SourceGit will call `git commit --file="TMP_MSG_FILE" <other params>` to finish commit

**Comment by love-linger** [comment]
I've push some changes to `develop` branch.

* Re-design the notification pannels
* Error message will not ignore blank lines.
* Rewrite the way reading commit full message while interactive rebasing.

**Comment by leo-liar** [comment]
> If your whitespace is in start of subject or endof description, it will be trimmed. Otherwise, all the content will be writed into a temp file, then SourceGit will call `git commit --file="TMP_MSG_FILE" <other params>` to finish commit

Thanks for the information! I could trace my problem to `git` itself (i.e. my missing whitespaces definitively aren't due to SourceGit's handling of the commit message.)

**Comment by Cyrillev91** [comment]
Hi,
Is it possible to add this point:
he 'Amend' checkbox should be automatically checked (by sourcegit software)  in the case of 'Edit' (interactive resabe) as explained #773

**Comment by love-linger** [comment]
> Hi, Is it possible to add this point: he 'Amend' checkbox should be automatically checked (by sourcegit software) in the case of 'Edit' (interactive resabe) as explained #773

I would like leave it unchecked. When a commit was marked as `edit` while interactively rebasing, user may just want to split the commit, not append changes.

**Comment by love-linger** [comment]
I'll close this issue since it contains some other things beyond title description. If you have more suggestions/questions, please open another issue
